sys/log/full/src/log_fcb2.c

```diff
@@ -0,0 +1,866 @@
+        return 0;
+    default:
+        return rc;
+    }
+
+#if MYNEWT_VAL(LOG_FCB_BOOKMARKS)
+    /* If a minimum index was specified (i.e., we are not just retrieving the
+     * last entry), add a bookmark pointing to this walk's start location.
+     */
+    if (log_off->lo_ts >= 0) {
+        fcb_log_add_bmark(fcb_log, &loc, log_off->lo_index);
+    }
+#endif
+
+    do {
+        rc = walk_func(log, log_off, &loc, loc.fe_data_len);
+        if (rc != 0) {
+            return rc;
+        }
+    } while (fcb_getnext(fcb, &loc) == 0);
+
+    return 0;
+}
+
+static int
+log_fcb2_flush(struct log *log)
+{
+    struct fcb_log *fcb_log;
+    struct fcb *fcb;
+
+    fcb_log = (struct fcb_log *)log->l_arg;
+    fcb = &fcb_log->fl_fcb;
+
+#if MYNEWT_VAL(LOG_FCB_BOOKMARKS)
+    fcb_log_clear_bmarks(fcb_log);
+#endif
+
+    return fcb_clear(fcb);
+}
+
+static int
+log_fcb2_registered(struct log *log)
+{
+    struct fcb *fcb;
+    int i;
+    struct fcb_log *fl;
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+#if MYNEWT_VAL(LOG_PERSIST_WATERMARK)
+    struct fcb_entry loc;
+#endif
+#endif
+
+    fl = (struct fcb_log *)log->l_arg;
+    fcb = &fl->fl_fcb;
+
+    for (i = 0; i < fcb->f_range_cnt; i++) {
+        if (fcb->f_ranges[i].fsr_align > LOG_FCB2_MAX_ALIGN) {
+            return SYS_ENOTSUP;
+        }
+    }
+
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+#if MYNEWT_VAL(LOG_PERSIST_WATERMARK)
+    /* Set watermark to first element */
+    memset(&loc, 0, sizeof(loc));
+
+    if (fcb_getnext(fcb, &loc)) {
+        fl->fl_watermark_off = loc.fe_area->fa_off + loc.fe_elem_off;
+    } else {
+        fl->fl_watermark_off = fcb->f_oldest->fa_off;
+    }
+#else
+    /* Initialize watermark to designated unknown value*/
+    fl->fl_watermark_off = 0xffffffff;
+#endif
+#endif
+    return 0;
+}
+
+#if MYNEWT_VAL(LOG_STORAGE_INFO)
+static int
+log_fcb2_storage_info(struct log *log, struct log_storage_info *info)
+{
+    struct fcb_log *fl;
+    struct fcb *fcb;
+    struct flash_area *fa;
+    uint32_t el_min;
+    uint32_t el_max;
+    uint32_t fa_min;
+    uint32_t fa_max;
+    uint32_t fa_size;
+    uint32_t fa_used;
+    int rc;
+
+    fl = (struct fcb_log *)log->l_arg;
+    fcb = &fl->fl_fcb;
+
+    rc = os_mutex_pend(&fcb->f_mtx, OS_WAIT_FOREVER);
+    if (rc && rc != OS_NOT_STARTED) {
+        return FCB_ERR_ARGS;
+    }
+
+    /*
+     * Calculate location of 1st entry.
+     * We assume 1st log entry starts at beginning of oldest sector in FCB.
+     * This is because even if 1st entry is in the middle of sector (is this
+     * even possible?) we will never use free space before it thus that space
+     * can be also considered used.
+     */
+    el_min = fcb->f_oldest->fa_off;
+
+    /* Calculate end location of last entry */
+    el_max = fcb->f_active.fe_area->fa_off + fcb->f_active.fe_elem_off;
+
+    /* Sectors assigned to FCB are guaranteed to be contiguous */
+    fa = &fcb->f_sectors[0];
```

Comment:
filed **f_sectors** not present in **FCB2** build

Reply:
Watermarking was not yet supported.